Add pytest plugin

gopiotr opened the pull request:
In this PR we added pytest plugin into Zephyr source code. It makes possibility to create and run more "advanced" tests, which require communication with flashed application (on hardware or simulator). Prepared shell test was made only to show basic pytest plugin functionality.

To test:
1. Run exemplary `shell` test on `native_posix` and `qemu`:
```
./scripts/twister -T samples/subsys/testsuite/pytest/shell -vv -p native_posix -p qemu_x86
```

2. Or run it on hardware:
```
./scripts/twister -T samples/subsys/testsuite/pytest/shell -vv -p nrf52840dk_nrf52840 --device-testing --device-serial /dev/ttyACM0
```

expected output (on `native_posix`):
```
DEBUG   - run test: native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell
DEBUG   - Running pytest command: export PYTHONPATH=/home/redbeard/zephyrproject/zephyr/scripts/pylib/pytest-twister-ext/src:${PYTHONPATH} && pytest --twister-ext -s -q /home/redbeard/zephyrproject/zephyr/samples/subsys/testsuite/pytest/shell/pytest --build-dir=/home/redbeard/zephyrproject/zephyr/twister-out/native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell --junit-xml=/home/redbeard/zephyrproject/zephyr/twister-out/native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/report.xml --log-level=DEBUG --device-type=native -p twister_ext.plugin
DEBUG   - PYTEST: 19:19:07.192:DEBUG: Get device type "native"
DEBUG   - PYTEST: 19:19:07.193:INFO: Running command: /home/redbeard/zephyrproject/zephyr/twister-out/native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/zephyr/zephyr.exe
DEBUG   - PYTEST: 19:19:07.194:DEBUG: Using selector: EpollSelector
DEBUG   - PYTEST: 19:19:07.197:DEBUG: Started subprocess with PID 160766
DEBUG   - PYTEST: 19:19:08.296:DEBUG: #: 
DEBUG   - PYTEST: 19:19:08.315:DEBUG: #: uart:~$ help
DEBUG   - PYTEST: 19:19:08.316:DEBUG: #: Please press the <Tab> button to see all available commands.
DEBUG   - PYTEST: .19:19:08.440:DEBUG: Stopping all running processes for PID 160766
DEBUG   - PYTEST: 19:19:08.454:INFO: Running simulation finished with return code 0
DEBUG   - PYTEST: 19:19:08.455:DEBUG: Get device type "native"
DEBUG   - PYTEST: 19:19:08.456:INFO: Running command: /home/redbeard/zephyrproject/zephyr/twister-out/native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/zephyr/zephyr.exe
DEBUG   - PYTEST: 19:19:08.456:DEBUG: Using selector: EpollSelector
DEBUG   - PYTEST: 19:19:08.459:DEBUG: Started subprocess with PID 160774
DEBUG   - PYTEST: 19:19:09.559:DEBUG: #: 
DEBUG   - PYTEST: 19:19:09.581:DEBUG: #: uart:~$ kernel version
DEBUG   - PYTEST: 19:19:09.581:DEBUG: #: Zephyr version 3.3.99
DEBUG   - PYTEST: .19:19:09.682:DEBUG: Stopping all running processes for PID 160774
DEBUG   - PYTEST: 19:19:09.688:INFO: Running simulation finished with return code 0
DEBUG   - PYTEST: - generated xml file: /home/redbeard/zephyrproject/zephyr/twister-out/native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/report.xml -
DEBUG   - PYTEST: 2 passed in 2.50s
DEBUG   - Pytest cases passed
DEBUG   - run status: native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell passed
INFO    - 1/1 native_posix              samples/subsys/testsuite/pytest/shell/sample.pytest.shell PASSED (native 2.510s)
```

---

**This PR consists of 3 parts:**

pytest-twister-ext plugin
---
New pytest plugin was added as a standalone python package into `scripts/pylib/pytest-twister-ext`.

Plugin has following structure:

`/scr/twister-ext/` - folder which consists all plugin source code, especially:
`/src/twister_ext/plugin.py` - file which contains plugin API (available options like `--build-dir`, `--device-type`, etc.).
`/scr/twister-ext/device/` - there are defined device adapter classes (dedicated for hardware, native_posix and QEMU).
`/scr/twister-ext/fixtures/` - at this moment there is placed one fixture `dut` which can be used in pytest test. It allow to automatically configure and flash/run application on device and return `DeviceAbstract` type object which can be used later in test for communication with device. There is plan to add there more advanced fixtures in the future.

`/tests/` - unit tests dedicated for plugin - at this moment the coverage rate is around 40% - there is plan to increase it in the future.

New pytest plugin was added as a standalone "python package", because it is required by pytest to make it possible to share fixtures, classes and methods among various tests. This plugin can be used without installing it - when Twister try to use this plugin at first it exports `PYTHONPATH` with path to this plugin, and then during call pytest it appends to command `-p twister_ext.plugin` argument. This allows to use this plugin installation it by pip.

Twister modifications
---

To make it possible to use effectively new pytest plugin it was necessary to make some modifications to the source code of Twister. When `harness: pytest` is chose in `sample.yaml` (or `testcase.yaml`) file, then Twister doesn't perform flashing hardware/running simulation actions, but instead of this it call pytest and pass to it all necessary information (like path to built application or port number of connected hardware). Then, flashing (or running) application is performed in pytest test by calling `dut` fixture. 

There is still possibility to run "old" pytest tests by Twister - (for example this one: https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/testsuite/pytest). The main difference is that now (with changes made in this PR) all pytest tests start with unflashed device. In entire Zephyr project we don't have hardware tests basing on pytest (beyond this one mentioned exemplary test), so it is hard to us to say how much this change could affect to some external project.

To summarize:
**If you already use Twister & pytest in your own project, and you expect to have flashed device before pytest test start, please add `dut` fixture call in your test function definition.**
If it's a big problem, we can consider to add something like `harness: pytest_old` option to maintain full backward compatibility.

Exemplary pytest shell test 
---
Old exemplary pytest test was move from `samples/subsys/testsuite/pytest/` to `samples/subsys/testsuite/pytest/basic` directory.
New exemplary pytest shell test was added to [samples/subsys/testsuite/pytest/shell](https://github.com/gopiotr/zephyr/tree/dev/pytest_plugin/samples/subsys/testsuite/pytest/shell). After call Twister by `./scripts/twister -T samples/subsys/testsuite/pytest/shell -vv -p native_posix`, basic shell application will be built, and then Twister will call pytest with all necessary options. At this moment there are defined two pytest tests in [test_shell.py](https://github.com/gopiotr/zephyr/blob/dev/pytest_plugin/samples/subsys/testsuite/pytest/shell/pytest/test_shell.py): `test_shell_print_help` and `test_shell_print_version`. Both tests use `dut` fixture (in their definition) and thanks to this connection with hardware/simulator is established automatically and then it can be used to communicate with device via methods like `write` or `iter_stdout`. First test sends to device `help` command and wait for expected output, second one sends `kernel version` command and wait for information about used kernel version.
All output received form device is saved to `$BUILD_DIR/handler.log` file (the same as it done by Twister for "regular" tests).

At this moment this test works properly on `native_posix`, QEMU (tested on `qemu_cortex_m3`, `qemu_cortex_a53`, `qemu_riscv32`, `qemu_leon3` and `qemu_x86`) and real hardware (tested on `nrf52840dk_nrf52840` and other nrf "family" boards).

---

Known issues
---

To make it possible to merge this PR into the main, at least following issues should be determined/added:

- [x] Determine how new pytest plugin should be installed ~~(indirectly via `scripts/requirements-run-test.txt` or directly by call `pip install ...`)~~. Now it is not necessary to install plugin by pip to use it and run pytest tests.
- [x] To make CI tests green - ~~update CI workflow files (`.github/workflows/twister.yaml` and `.github/workflows/clang.yaml`) to make it possible to install new pytest plugin and thus make it available in pytest tests which currently fail~~. Now it is not necessary to install plugin by pip to use it and run pytest tests, so `.github/workflows/*.yaml` don't need to be updated.
- [x] Add minimal documentation to describe how new pytest plugin works and how it can be used in tests.
- [x] ~~Update `CODEOWNERS` and `MAINTAINERS.yml` files.~~ https://github.com/zephyrproject-rtos/zephyr/pull/57117#issuecomment-1562682017
- [x] Add to `.github/workflows/twister_tests.yml` CI workflow unit tests dedicated to new pytest plugin.

Other known issues which will be added/developed in the next PRs, when this PR will be merged:

- ~At this moment Twister options like `--pre-script`, `--post-flash-script` and `--post-script` are not handled by pytest plugin during flashing/running application. We decided that this functionality is not "fundamental" and could be added with next PRs.~
- It should be determined if some part of code could be shared somehow between Twister and pytest plugin to avoid code duplication.
- Currently when Twister call pytest, then it reports in final report (xml or json) whole pytest call as one test. It would be nice to make it possible to add to final report also pytest testcases (the same as Ztest testcases are added).
- At this moment device adapters in pytest plugin provide `iter_stdout` method to read something from device. In some cases it is not the most convenient way, and it should be determined how to do it in better way (for example replace it by simple `read` function with byte numbers and timeout arguments).

Tracking issue: https://github.com/zephyrproject-rtos/zephyr/issues/58288

---

This PR don't focus on testing shell applications, but provides support for such a kind of tests, so I attach following issue:
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/52889

This PR improves integration with pytest and thanks to this it should be easier to write tests, which communicate with external programs:
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/27956

nashif commented:
does not seem to work with qemu...

getting:
```
Can't access report.xml
Pytest cases failed

Output from pytest:
Traceback (most recent call last):
  File "/home/nashif/.local/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 187, in console_main
    code = main()
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 145, in main
    config = _prepareconfig(args, plugins)
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 324, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
  File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_callers.py", line 55, in _multicall
    gen.send(outcome)
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/helpconfig.py", line 102, in pytest_cmdline_parse
    config: Config = outcome.get_result()
  File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1016, in pytest_cmdline_parse
    self.parse(args)
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1304, in parse
    self._preparse(args, addopts=addopts)
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1190, in _preparse
    self.known_args_namespace = self._parser.parse_known_args(
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/argparsing.py", line 152, in parse_known_args
    return self.parse_known_and_unknown_args(args, namespace=namespace)[0]
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/argparsing.py", line 161, in parse_known_and_unknown_args
    optparser = self._getparser()
  File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/argparsing.py", line 128, in _getparser
    arggroup.add_argument(*n, **a)
  File "/usr/lib/python3.10/argparse.py", line 1453, in add_argument
    return self._add_action(action)
  File "/usr/lib/python3.10/argparse.py", line 1655, in _add_action
    action = super(_ArgumentGroup, self)._add_action(action)
  File "/usr/lib/python3.10/argparse.py", line 1467, in _add_action
    self._check_conflict(action)
  File "/usr/lib/python3.10/argparse.py", line 1604, in _check_conflict
    conflict_handler(action, confl_optionals)
  File "/usr/lib/python3.10/argparse.py", line 1613, in _handle_conflict_error
    raise ArgumentError(action, message % conflict_string)
argparse.ArgumentError: argument -O/--outdir: conflicting option strings: -O, --outdir
```

gopiotr commented:
> does not seem to work with qemu...

@nashif yes, I know about it - this is initial version of final PR, and there is a plan to add handling for QEMU and other hardware. At this moment it works only with big limitations. I wrote better PR description to avoid confusion. I'll inform you when it will be ready for review ;)

gchwier commented:
Added support for `qemu` and `native_posix`.
A new test sample can be found in `samples/subsys/testsuite/pytest/shell`.

To install / update pytest-twister-ext plugin, use:
`pip install -e scripts/pylib/pytest-twister-ext`

One can test plugin with commands:
run qemu and native_posix
`./scripts/twister -T samples/subsys/testsuite/pytest/shell -vv -p native_posix -p qemu_x86`

run on device with west-flash options:
`./scripts/twister -T samples/subsys/testsuite/pytest/shell -vv -p nrf9160dk_nrf9160 --device-testing --device-serial /dev/ttyACM0 --west-flash='--erase,--dev-id=xxxxxxx'

or with hardware map:
`./scripts/twister -T samples/subsys/testsuite/pytest/shell -vv --device-testing --hardware-map ~/tmp/hardware_map.yml`

also added unit tests of pytest-twister-ext plugin, to run it use:
`pytest -s -vv scripts/pylib/pytest-twister-ext/tests`

teburd commented on scripts/requirements-run-test.txt:
```diff
@@ -13,3 +13,6 @@ cbor>=1.0.0
 psutil
+
+# pytest plugin
+-e ${ZEPHYR_BASE}/scripts/pylib/pytest-twister-ext
```
pip install -r scripts/requirements-run-test.txt fails for me with this line

gchwier replied:
`ZEPHYR_BASE=<path to Zephyr dir>` must be exported when installing `pytest-twister-ext` from requirements file. Or just install it manually with `pip install -e scripts/pylib/pytest-twister-ext`

teburd reviewed (commented):
twister -p qemu_x86_64 -T samples/subsys/testsuite/pytest

~Fails for me without any cause, maybe I'm missing something important but this just doesn't seem to work for me.~

Ok works fine after pip install -e scripts/pylib/pytest-twister-ext

dreiss commented on scripts/pylib/pytest-twister-ext/src/twister_ext/device/device_abstract.py:
```diff
@@ -0,0 +1,94 @@
+    def stop(self) -> None:
+        """Stop device."""
+
+    # @abc.abstractmethod
+    # def read(self, size=1) -> None:
+    #     """Read size bytes from device"""
+
+    # def read_until(self, expected, size=None):
+    #     """Read until an expected bytes sequence is found"""
+    #     lenterm = len(expected)
+    #     line = bytearray()
+    #     while True:
+    #         c = self.read(1)
+    #         if c:
+    #             line += c
+    #             if line[-lenterm:] == expected:
+    #                 break
+    #             if size is not None and len(line) >= size:
+    #                 break
+    #         else:
+    #             break
+    #     return bytes(line)
```
Would be great to see these enabled.  For tests that need back-and-forth communication with the device, iter_stdout might not be expressive enough.

gchwier replied:
Yes, I agree. We left that part of code (commented), to enable it in next iteration. It works for hardware and qemu, but there were problems in `simulator_adapter.py` - used for `native_posix` (we think about changing implementation to not use `async` there)

dreiss commented on scripts/pylib/pytest-twister-ext/src/twister_ext/device/qemu_adapter.py:
```diff
@@ -0,0 +1,196 @@
+
+    @property
+    def iter_stdout(self) -> Generator[str, None, None]:
+        if not self.connection:
+            return
+        # fifo file can be not create yet, so we need to wait for a while
+        self._wait_for_fifo()
+
+        # create unblocking reading from fifo file
+        q: Queue = Queue()
+
+        def read_lines():
+            while self.connection and self.connection.is_open:
+                try:
+                    line = self.connection.readline().decode('UTF-8').strip()
+                except (OSError, ValueError):
+                    # file could be closed already so we should stop reading
+                    break
+                if len(line) != 0:
+                    q.put(line)
+
+        t = threading.Thread(target=read_lines, daemon=True)
+        t.start()
+
+        end_time = time.time() + self.timeout
+        try:
+            while True:
+                try:
+                    stream = q.get(timeout=0.1)
+                    self.handler_log_file.handle(data=stream + '\n')
+                    yield stream
+                except Empty:  # timeout appeared
+                    pass
+                if time.time() > end_time:
+                    break
+        except KeyboardInterrupt:
+            # let thread to finish smoothly
+            pass
+        finally:
+            t.join(1)
```
Another option would be to have QEMU connect its UARTs to Unix-domain sockets.  Here's a snippet I did for a test I was working on: https://gist.github.com/dreiss/63fed74aa7f43527a987916b646fc860 .  This scales to multiple UARTs and doesn't require a separate thread as long as you're draining the buffers frequently enough.

dreiss commented on scripts/pylib/pytest-twister-ext/src/twister_ext/device/qemu_adapter.py:
```diff
@@ -0,0 +1,196 @@
+        self.timeout: float = 60  # running timeout in seconds
+        self.booting_timeout_in_ms: int = 10_000  #: wait time for booting Qemu in milliseconds
+
+    def generate_command(self) -> None:
```
Would be great to have an option to enable the QEMU gdb server as well.

teburd commented:
I already have a use case for this, and I think its good enough to be usable as is so long as the reporting is working, :+1: from me

nashif commented:
still getting this:

```
EBUG   - PYTEST: argparse.ArgumentError: argument -O/--outdir: conflicting option strings: -O, --outdir
WARNING - Can't access report.xml
INFO    - Pytest cases failed.
DEBUG   - run status: native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell failed
INFO    - 1/2 native_posix              samples/subsys/testsuite/pytest/shell/sample.pytest.shell  FAILED Pytest failed (native)
ERROR   - see: /home/nashif/zephyrproject/zephyr/twister-out/native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/build.log
DEBUG   - PYTEST: Traceback (most recent call last):
DEBUG   - PYTEST: File "/home/nashif/.local/bin/pytest", line 8, in <module>
DEBUG   - PYTEST: sys.exit(console_main())
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 189, in console_main
DEBUG   - PYTEST: code = main()
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 147, in main
DEBUG   - PYTEST: config = _prepareconfig(args, plugins)
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 328, in _prepareconfig
DEBUG   - PYTEST: config = pluginmanager.hook.pytest_cmdline_parse(
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_hooks.py", line 265, in __call__
DEBUG   - PYTEST: return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_manager.py", line 80, in _hookexec
DEBUG   - PYTEST: return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_callers.py", line 55, in _multicall
DEBUG   - PYTEST: gen.send(outcome)
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/helpconfig.py", line 103, in pytest_cmdline_parse
DEBUG   - PYTEST: config: Config = outcome.get_result()
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_result.py", line 60, in get_result
DEBUG   - PYTEST: raise ex[1].with_traceback(ex[2])
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/pluggy/_callers.py", line 39, in _multicall
DEBUG   - PYTEST: res = hook_impl.function(*args)
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1067, in pytest_cmdline_parse
DEBUG   - PYTEST: self.parse(args)
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1354, in parse
DEBUG   - PYTEST: self._preparse(args, addopts=addopts)
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1240, in _preparse
DEBUG   - PYTEST: self.known_args_namespace = self._parser.parse_known_args(
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/argparsing.py", line 157, in parse_known_args
DEBUG   - PYTEST: return self.parse_known_and_unknown_args(args, namespace=namespace)[0]
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/argparsing.py", line 171, in parse_known_and_unknown_args
DEBUG   - PYTEST: optparser = self._getparser()
DEBUG   - PYTEST: File "/home/nashif/.local/lib/python3.10/site-packages/_pytest/config/argparsing.py", line 130, in _getparser
DEBUG   - PYTEST: arggroup.add_argument(*n, **a)
DEBUG   - PYTEST: File "/usr/lib/python3.10/argparse.py", line 1453, in add_argument
DEBUG   - PYTEST: return self._add_action(action)
DEBUG   - PYTEST: File "/usr/lib/python3.10/argparse.py", line 1655, in _add_action
DEBUG   - PYTEST: action = super(_ArgumentGroup, self)._add_action(action)
DEBUG   - PYTEST: File "/usr/lib/python3.10/argparse.py", line 1467, in _add_action
DEBUG   - PYTEST: self._check_conflict(action)
DEBUG   - PYTEST: File "/usr/lib/python3.10/argparse.py", line 1604, in _check_conflict
DEBUG   - PYTEST: conflict_handler(action, confl_optionals)
DEBUG   - PYTEST: File "/usr/lib/python3.10/argparse.py", line 1613, in _handle_conflict_error
DEBUG   - PYTEST: raise ArgumentError(action, message % conflict_string)
DEBUG   - PYTEST: argparse.ArgumentError: argument -O/--outdir: conflicting option strings: -O, --outdir
WARNING - Can't access report.xml
INFO    - Pytest cases failed.
DEBUG   - run status: qemu_x86/samples/subsys/testsuite/pytest/shell/sample.pytest.shell failed
INFO    - 2/2 qemu_x86                  samples/subsys/testsuite/pytest/shell/sample.pytest.shell  FAILED Pytest failed (qemu)
ERROR   - see: /home/nashif/zephyrproject/zephyr/twister-out/qemu_x86/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/build.log

INFO    - 1 test scenarios (2 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
```

gchwier commented:
@nashif you have probably installed the TwisterV2 package (name of package is `pytest-twister`) in your env. You should uninstall it: `pip uninstall pytest-twister`. In both packages (old twisterV2 and current pytest-twister-ext) are same keywords used with argparse, and when pytest loads packages, there is a conflict.

gopiotr commented:
I updated the PR description. I provided there more information on what this PR consists of and what "known issues" are (with explanation why CI tests fail).

nashif commented:
looks good sofar, my only issue is the fact we have to install a pip module to get this working. How exactly is this going to work when bisecting for example, or jumping between branches, etc? How do we keep the the in-tree version of twister compatible with this twister extension? Do I have to reinstall the module everytime I move between revisions?

gchwier commented:
@nashif When you are using `-e, --editable` mode, then installed package is linked to the provided location. So you don't need to reinstall the module when you do any change (e.g changing the branch)

nashif commented:
> @nashif When you are using -e, --editable mode, then installed package is linked to the provided location. So you don't need to reinstall the module when you do any change (e.g changing the branch)

ok, that sounds ok, unless you are someone like me who tend to always keep their tree clean and very often calls `git clean -dxf`, so, I have to install the module everytime I call twister after a tree cleanup?

gopiotr commented:
@nashif to make it works installing package with `--editable` mode and still use `git clean -dxf` command you can try following steps:
1. Uninstall current version of `pytest-twister-ext`:
```
pip uninstall pytest-twister-ext
```
2. Upgrade your `setuptools` package (I work with `67.7.2` version)
```
pip install --upgrade setuptools
```
3. If you have already installed `wheel` package (which is not required in Zephyr project), please uninstall it:
```
pip uninstall wheel
```
4. Now, install once again `pytest-twister-ext` plugin with `--editable` mode:
```
pip install --editable scripts/pylib/pytest-twister-ext/
```
5. Apply `git clean -dxf` into Zephyr project:
```
git clean -dxf
```
6. Try to run pytest tests by command:
```
./scripts/twister -T samples/subsys/testsuite/pytest/shell -vv -p native_posix
```
It works for me, and I'am able to apply `git clean -dxf` multiple times, switch among various branches, and when I switch back to this PR, I can call exemplary pytest shell test from this PR and it pass.

It seems, that there were changes in `setuptools` and newer versions do not require `.egg-info` file, which is created when `pip install ...` command is called. If I'll find what is minimal required version for `setuptools` I could add it to one of `requirements.txt` files.

Please let me know how it works in your environment.

gopiotr commented:
> @nashif to make it works installing package with `--editable` mode and still use `git clean -dxf` command you can try following steps:
> 
> ...

I added `pyproject.toml` file with specified `build-backend = "setuptools.build_meta"` and now my proposition works even with installed `wheel` package.

nashif commented:
now getting this weird error, have never seen this before:

```
      Successfully uninstalled grpcio-1.54.0
  Running setup.py develop for pytest-twister-ext
    error: subprocess-exited-with-error

    × python setup.py develop did not run successfully.
    │ exit code: 1
    ╰─> [32 lines of output]
        running develop
        /usr/lib/python3/dist-packages/setuptools/command/easy_install.py:158: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
          warnings.warn(
        WARNING: The user site-packages directory is disabled.
        /usr/lib/python3/dist-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
          warnings.warn(
        error: can't create or remove files in install directory

        The following error occurred while trying to add or remove files in the
        installation directory:

            [Errno 13] Permission denied: '/usr/local/lib/python3.10/dist-packages/test-easy-install-3166548.write-test'

        The installation directory you specified (via --install-dir, --prefix, or
        the distutils default setting) was:

            /usr/local/lib/python3.10/dist-packages/

        Perhaps your account does not have write access to this directory?  If the
        installation directory is a system-owned directory, you may need to sign in
        as the administrator or "root" account.  If you do not have administrative
        access to this machine, you may wish to choose a different installation
        directory, preferably one that is listed in your PYTHONPATH environment
        variable.

        For information on other options, you may wish to consult the
        documentation at:

          https://setuptools.pypa.io/en/latest/deprecated/easy_install.html

        Please make the appropriate changes for your system and try again.

        [end of output]

    note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× python setup.py develop did not run successfully.
│ exit code: 1
╰─> [32 lines of output]
    running develop
    /usr/lib/python3/dist-packages/setuptools/command/easy_install.py:158: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
      warnings.warn(
    WARNING: The user site-packages directory is disabled.
    /usr/lib/python3/dist-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
      warnings.warn(
    error: can't create or remove files in install directory

    The following error occurred while trying to add or remove files in the
    installation directory:

        [Errno 13] Permission denied: '/usr/local/lib/python3.10/dist-packages/test-easy-install-3166548.write-test'

    The installation directory you specified (via --install-dir, --prefix, or
    the distutils default setting) was:

        /usr/local/lib/python3.10/dist-packages/

    Perhaps your account does not have write access to this directory?  If the
    installation directory is a system-owned directory, you may need to sign in
    as the administrator or "root" account.  If you do not have administrative
    access to this machine, you may wish to choose a different installation
    directory, preferably one that is listed in your PYTHONPATH environment
    variable.

    For information on other options, you may wish to consult the
    documentation at:

      https://setuptools.pypa.io/en/latest/deprecated/easy_install.html

    Please make the appropriate changes for your system and try again.

    [end of output]

```

gopiotr commented:
> now getting this weird error, have never seen this before:

@nashif we changed a little bit our strategy. We added possibility of use our plugin in Twister without installing it by `pip`. So, now please uninstall `pytest-twister-ext` by:
```
pip uninstall pytest-twister-ext
```
and clean any "leftovers" in plugin source code by:
```
git clean -dxf scripts/pylib/pytest-twister-ext/
```
After this you should be able to run exemplary shell test without installed plugin, by just:
```
./scripts/twister -T samples/subsys/testsuite/pytest/shell/ -vv -p native_posix
```
Then you can observe, that pytest command called by Twister was changed and now it looks like this:
```
export PYTHONPATH=/home/redbeard/zephyrproject/zephyr/scripts/pylib/pytest-twister-ext/src:${PYTHONPATH} && pytest --twister-ext -s -q /home/redbeard/zephyrproject/zephyr/samples/subsys/testsuite/pytest/shell/pytest --build-dir=/home/redbeard/zephyrproject/zephyr/twister-out/native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell --junit-xml=/home/redbeard/zephyrproject/zephyr/twister-out/native_posix/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/report.xml --log-level=DEBUG --device-type=native -p twister_ext.plugin
```
By exporting `PYTHONPATH` variable and adding `-p twister_ext.plugin` argument to pytest command, now it is possible to run pytest tests without plugin installation, what solve problems connected with installing it with editable mode, and cleaning repository by `git clean -dxf`.

Please let us know, what do you think about this approach?

PerMac commented:
@nashif From our research, it is impossible to have the pytest plugin registered twice (when installed with pip and sources injected with PYTHONPATH). There is no way to tell which one to use (order doesn't matter since both are discovered). We propose 2 options to fix this:
- add a warning `You work with installed version of pytest-twister-harness plugin - make sure that you work with the proper version`, printed when the installed version is discovered and continue to run
- raise error and exit when the installed version is discovered. Add flag "--allow-installed-plugin" disabling the error.
Which one do you prefer?

Contex: During latest testing wg meeting Anas raised a conceren that (paraphrazing)  "since plugin is installable (even if not mentioned in instruction/docs) one can install it and then be surprised (or worse, not noticing) that  modifications in source code are not applied (e.g. from checkouted branches). Such issue cannot take place."

PerMac commented:
We started from the plugin approach but were convinced that it should be developed in tree. I think some reasonings are given here: https://github.com/zephyrproject-rtos/zephyr/pull/57117#issuecomment-1542123542  I realized most of the concerns why it shouldn't be a real plugin were only verbal during our testing meeting. @nashif can you please write your reasoning here?

nashif commented:
> We started from the plugin approach but were convinced that it should be developed in tree. I think some reasonings are given here: [#57117 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/57117#issuecomment-1542123542) I realized most of the concerns why it shouldn't be a real plugin were only verbal during our testing meeting. @nashif can you please write your reasoning here?

Before any code was submitted, we agreed that this functionality should be implemented to be part of twister and the zephyr tree. There was a desire to maintain this out of tre and I was fully opposed to this given how twister is maintained and how it is being developed right now. See comment mentioned above for why this will be a problem and how it creates fragmentation and complexities in both using and developing twister.
If we want to move to an out-of-tree twister with plugin support it needs to be done all the way and not for some feature because it is easier for 1 developer or use-case and we have to address the issues of bisectability, branches and so on. 

I also completely disagree with having this particular code have its own coding guidelines, CI and syntax checking. If  we need to do that, it has to be done for the complete twister code base. I do not see this features as something isolated that will live on its own and have its own roadmap. As I said before, I see this becoming one of the most use ways of testing and we will have 100s of tested converted or implemented to use this feature, so it is going a core twister testing harness and has to be compatible and following the same rules we use for twister development.

I have many review comments lined up for this PR that will come in later, but this is a summary until then.

hakehuang reviewed (commented):
can we have some where to document the pytest supporting features? as pytest is on tree?

nashif reviewed (changes requested):
few comments while I go deeper in the code and documenting some feedback given in the testing WG meeting.

gopiotr commented on scripts/pylib/pytest-device-adapter/.pre-commit-config.yaml:
```diff
@@ -0,0 +1,36 @@
+repos:
```
> how is this file going to be used exactly?

If you are interested in how `pre-commit` works, you can find more information here: https://pre-commit.com/. We used it a lot during our work and it was very helpful. In general we decided to use more linters (like `flake8`, `isort` or `mypy`), because that helped us to maintain our code cleaner. In our initial idea we thought that it could be valuable to use them also in the further development and in CI pipelines - that's why we placed in this PR `.pre-commit-config.yaml` and `tox.ini` files. We don't have enough resources now to change and apply coding guidelines in the whole Twister project. So, if it is unacceptable to you to have different one for this pytest plugin, then I'll remove all those extra files from this PR.

gopiotr commented on scripts/pylib/pytest-device-adapter/README.rst:
```diff
@@ -0,0 +1,37 @@
+
+.. code-block:: sh
+
+  pip install .
```
I believe that it can be helpful for developers who work more with pytest tests (like me) to have it as an installed package. But if it's so confusing for you, I remove this instruction from the `README` file.

gopiotr commented on scripts/pylib/pytest-device-adapter/requirements.txt:
```diff
@@ -0,0 +1,5 @@
+marshmallow==3.19.0
```
Okay, removed.

PerMac commented:
@hake Docs added

PerMac commented:
@nashif Do you think it can enter the main before 3.4 release? Can we help you somehow to make it possible?
We plan to squash the commits before merging, leaving them for now to help with reviewing the changes

nashif commented on doc/develop/test/pytest.rst:
```diff
@@ -0,0 +1,84 @@
+Integration with pytest test framework
+######################################
+
+*Please mind that integration of twister with pytest is still work in progress. Not every platform
```
Can you please open a tacking issue for all things implemented and remaining items (preferably as a task list) to have an overview of what is done and what is coming?

nashif commented on doc/develop/test/pytest.rst:
```diff
@@ -0,0 +1,84 @@
+Limitations
+***********
+
+* Twister options like ``--pre-script``, ``--post-flash-script`` and ``--post-script`` are not handled by
```
why not? is this planned? can you please put this in the issue mentioned above?

nashif replied:
btw, this excludes majority of boards and devices from using pytest, so this is a blocker issue for supporting devices.

PerMac replied:
Can we add this to the "tracking issue"? We looked at this and adding this seems rather straight forward and can be done in a next iteration.

nashif replied:
why can't this be supported by passing the pre/post scripts as command line options to pytest like everything else and call them from pytest? It should be easy enough to do, you are just calling a script...

nashif commented on doc/develop/test/pytest.rst:
```diff
@@ -0,0 +1,84 @@
+
+* Twister options like ``--pre-script``, ``--post-flash-script`` and ``--post-script`` are not handled by
+  pytest plugin during flashing/running application.
+* The whole pytest call is reported as one test in the final twister report (xml or json).
```
ditto

nashif commented on doc/develop/test/pytest.rst:
```diff
@@ -0,0 +1,84 @@
+* Twister options like ``--pre-script``, ``--post-flash-script`` and ``--post-script`` are not handled by
+  pytest plugin during flashing/running application.
+* The whole pytest call is reported as one test in the final twister report (xml or json).
+* Device adapters in pytest plugin provide `iter_stdout` method to read from devices. In some
```
ditto

nashif commented on doc/develop/test/pytest.rst:
```diff
@@ -0,0 +1,84 @@
+* Device adapters in pytest plugin provide `iter_stdout` method to read from devices. In some
+  cases, it is not the most convenient way, and it will be considered how to improve this
+  (for example replace it with a simple read function with a given byte size and timeout arguments).
+* Not every platform type is supported in the plugin (yet).
```
ditto

nashif commented on scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py:
```diff
@@ -0,0 +1,202 @@
+        if runner := self.device_config.runner:
+            command.extend(['--runner', runner])
+
+            if board_id := self.device_config.id:
```
as discussed already, this is a major flaw if we have to maintain such code twice, we need to find a better and only 1 place to implement this logic so there will not be any need to change this code here whenever we add a new flasher or platform specific tooling.

So please add this as something to be resolved in the tracking issue.

nashif commented on scripts/pylib/pytest-twister-harness/src/twister_harness/device/qemu_adapter.py:
```diff
@@ -0,0 +1,196 @@
+logger = logging.getLogger(__name__)
+
+
+class QemuAdapter(DeviceAbstract):
```
same here, lots of duplication. We need to find a better way to do that and only in one place.

nashif commented on scripts/pylib/pytest-twister-harness/src/twister_harness/plugin.py:
```diff
@@ -0,0 +1,136 @@
+)
+
+
+def pytest_addoption(parser: pytest.Parser):
```
Using command line options as the interface between twister core and the harness plugin is not great, this is error prone and depending on  the target and what options it needs for serial capturing and flashing might get too complex and too long. We need to look into a better way of interfacing with the harness other than command line options. Please add this as a future improvement in the tracking issue.

nashif commented on scripts/pylib/pytest-twister-harness/tests/device/data/fifo_mock.py:
```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2023 Nordic Semiconductor ASA
+#
```
when is this test being run?

PerMac replied:
added to the "Twister TestSuite / Twister Unit Tests" check (workflow twister_tests.yaml) and is already passing on this PR.

nashif commented on scripts/pylib/pytest-twister-harness/tests/device/data/mock_script.py:
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/python
+# Copyright (c) 2023 Nordic Semiconductor ASA
```
who runs this test and when?

PerMac replied:
added to the "Twister TestSuite / Twister Unit Tests" check (workflow twister_tests.yaml) and is already passing on this PR.

nashif reviewed (commented):
few more comments, main ask is to create a tracking issue of all upcoming changes and plans per the discussions in the testing WG.

nashif commented:
> Update CODEOWNERS and MAINTAINERS.yml files. ❓

This is part of twister, no need to create a new entry in MAINTAINERS file

> Add to .github/workflows/twister_tests.yml CI workflow unit tests dedicated to new pytest plugin. ❓

Yes, we need those to run as part of the twiser tests workflow.

PerMac commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -765,6 +774,16 @@ def parse_arguments(parser, args, options = None):
         options.extra_test_args = options.extra_test_args[1:]
 
+    if not options.allow_installed_plugin and PYTEST_PLUGIN_INSTALLED:
+        logger.error("By default Twister should work with uninstalled "
```
changed as you've suggested

PerMac commented:
@nashif Thanks for the comments!
I see there are 2 major topics to address (or did I miss something?):
- issue where we track the current status and things to be resolved: will open it today
- unit tests (who runs this test and when?): those are meant to be run be the twister_test workflow (where other twister's unit tests are called), the latest commit adds those tests to twister_test.yaml. Since it is "on: pull_request" (without _target) the change is already visible and those tests are executed in this PR

nashif commented:
> or did I miss something?

the pre/post scripts would be the other one. It was mentiioned it should be easy, and I think it is. So let's get this in.

PerMac commented:
The latest commit adds support for pre/post-flash/post -scripts. @nashif @evgeniy-paltsev Can you please give it a spin?

evgeniy-paltsev commented:
> @evgeniy-paltsev Can you please give it a spin on your boards?

Will do.

BTW, I'm wondering if this has same requirements about python version as Zephyr itself (python >= 3.8)?

Or we need something newer?

PerMac commented:
> > @evgeniy-paltsev Can you please give it a spin on your boards?
> 
> Will do.
> 
> BTW, I'm wondering if this has same requirements about python version as Zephyr itself (python >= 3.8)?
> 
> Or we need something newer?

It has the same requirements. You don't have to do anything to get it activated (if you have this PR checked out), as stated in docs and readme. It is even already executed in this CI without any modification of the plan. Also the plan with pytest unit tests for twister itself is expanded in this PR adding tests for pytest-twister-plugin. This plan is tested against python 3.8, 3.9 and 3.10 and all pass.

PerMac commented:
@nashif Tracking issue: https://github.com/zephyrproject-rtos/zephyr/issues/58288

nashif commented:
It appears that we are flashing twice? Something is wrong here. I see the device being flashed twice by observing the device LEDs and it is seen here in the log as well.

```
DEBUG   - Running pytest command: export PYTHONPATH=/home/nashif/zephyrproject/zephyr/scripts/pylib/pytest-twister-harness/src:${PYTHONPATH} && pytest --twister-harness -s -q /home/nashif/zephyrproject/zephyr/samples/subsys/testsuite/pytest/shell/pytest --build-dir=/home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell --junit-xml=/home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/report.xml --log-level=DEBUG --device-type=hardware --device-serial=/dev/ttyACM0 --device-serial-baud=115200 --runner=pyocd --device-id=0240000026334e450015400f5e0e000b4eb1000097969900 '--device-product=DAPLink CMSIS-DAP' -p twister_harness.plugin
DEBUG   - PYTEST: 20:44:07.70:DEBUG: Get device type "hardware"
DEBUG   - PYTEST: 20:44:07.70:INFO: Opening serial connection for /dev/ttyACM0
DEBUG   - PYTEST: 20:44:07.70:INFO: Flashing device 0240000026334e450015400f5e0e000b4eb1000097969900
DEBUG   - PYTEST: 20:44:07.71:INFO: Flashing command: /home/nashif/.local/bin/west flash --skip-rebuild --build-dir /home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell --runner pyocd -- --board-id 0240000026334e450015400f5e0e000b4eb1000097969900
DEBUG   - PYTEST: 20:44:10.552:DEBUG: -- west flash: using runner pyocd
DEBUG   - PYTEST: WARNING: runners.pyocd: hex file (None) does not exist; falling back on .bin (/home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/zephyr/zephyr.bin). Consider enabling CONFIG_BUILD_OUTPUT_HEX.
DEBUG   - PYTEST: -- runners.pyocd: Flashing file: /home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/zephyr/zephyr.bin
DEBUG   - PYTEST: 0001017 I Loading /home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/zephyr/zephyr.bin [load_cmd]
DEBUG   - PYTEST: [---|---|---|---|---|---|---|---|---|----]
DEBUG   - PYTEST: [========================================]
DEBUG   - PYTEST: 0002900 I Erased 0 bytes (0 sectors), programmed 0 bytes (0 pages), skipped 57344 bytes (14 pages) at 29.78 kB/s [loader]
DEBUG   - PYTEST: 20:44:10.552:INFO: Flashing finished
DEBUG   - PYTEST: 20:44:11.556:DEBUG: #: help
DEBUG   - PYTEST: 20:44:11.561:DEBUG: #: Please press the <Tab> button to see all available commands.
DEBUG   - PYTEST: 20:44:11.569:DEBUG: #: You can also use the <Tab> button to prompt or auto-complete all commands or its subcommands.
DEBUG   - PYTEST: 20:44:11.577:DEBUG: #: You can try to call commands with <-h> or <--help> parameter for more information.
DEBUG   - PYTEST: 20:44:11.580:DEBUG: #: Shell supports following meta-keys:
DEBUG   - PYTEST: 20:44:11.583:DEBUG: #: Ctrl + (a key from: abcdefklnpuw)
DEBUG   - PYTEST: 20:44:11.585:DEBUG: #: Alt  + (a key from: bf)
DEBUG   - PYTEST: 20:44:11.590:DEBUG: #: Please refer to shell documentation for more details.
DEBUG   - PYTEST: 20:44:11.592:DEBUG: #: Available commands:
DEBUG   - PYTEST: .20:44:11.613:INFO: Closed serial connection for /dev/ttyACM0
DEBUG   - PYTEST: 20:44:11.615:DEBUG: Get device type "hardware"
DEBUG   - PYTEST: 20:44:11.615:INFO: Opening serial connection for /dev/ttyACM0
DEBUG   - PYTEST: 20:44:11.616:INFO: Flashing device 0240000026334e450015400f5e0e000b4eb1000097969900
DEBUG   - PYTEST: 20:44:11.616:INFO: Flashing command: /home/nashif/.local/bin/west flash --skip-rebuild --build-dir /home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell --runner pyocd -- --board-id 0240000026334e450015400f5e0e000b4eb1000097969900
DEBUG   - PYTEST: 20:44:15.118:DEBUG: -- west flash: using runner pyocd
DEBUG   - PYTEST: WARNING: runners.pyocd: hex file (None) does not exist; falling back on .bin (/home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/zephyr/zephyr.bin). Consider enabling CONFIG_BUILD_OUTPUT_HEX.
DEBUG   - PYTEST: -- runners.pyocd: Flashing file: /home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/zephyr/zephyr.bin
DEBUG   - PYTEST: 0001003 I Loading /home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/zephyr/zephyr.bin [load_cmd]
DEBUG   - PYTEST: [---|---|---|---|---|---|---|---|---|----]
DEBUG   - PYTEST: [========================================]
DEBUG   - PYTEST: 0002889 I Erased 0 bytes (0 sectors), programmed 0 bytes (0 pages), skipped 57344 bytes (14 pages) at 29.74 kB/s [loader]
DEBUG   - PYTEST: 20:44:15.118:INFO: Flashing finished
DEBUG   - PYTEST: 20:44:16.123:DEBUG: #: kernel version
DEBUG   - PYTEST: 20:44:16.125:DEBUG: #: Zephyr version 3.3.99
DEBUG   - PYTEST: .20:44:16.133:INFO: Closed serial connection for /dev/ttyACM0
DEBUG   - PYTEST: - generated xml file: /home/nashif/zephyrproject/zephyr/twister-out/frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell/report.xml -
DEBUG   - PYTEST: 2 passed in 9.07s
DEBUG   - Could not find a matching testcase for sample.pytest.shell
DEBUG   - Pytest cases passed
DEBUG   - run status: frdm_k64f/samples/subsys/testsuite/pytest/shell/sample.pytest.shell passed
INFO    - 1/1 frdm_k64f                 samples/subsys/testsuite/pytest/shell/sample.pytest.shell PASSED (device 9.074s)

```

gchwier commented:
> 

There are two tests in https://github.com/gopiotr/zephyr/blob/dev/pytest_plugin/samples/subsys/testsuite/pytest/shell/pytest/test_shell.py
and there is flashing before every test.

PerMac commented:
@nashif However, the configuration is built only once and the image is reused. The dut fixture has a flashing in its setup.  It is to support the test isolation principle.  It is always possible to have more asserts in a single test if you don't want reflashing

gopiotr committed:
scripts: add pytest plugin

Adding pytest plugin dedicated to running pytest tests in Zephyr
project. This plugin provides a dut fixture which allows to handle
bidirectional communication with the device under test. This version
of plugin can be used for tests dedicated to real hardware, QEMU and
native_posix simulator.

Co-authored-by: Lukasz Fundakowski <lukasz.fundakowski@nordicsemi.no>
Co-authored-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>
Co-authored-by: Katarzyna Giadla <katarzyna.giadla@nordicsemi.no>
Signed-off-by: Piotr Golyzniak <piotr.golyzniak@nordicsemi.no>

gchwier committed:
scripts: twister: adaptation for pytest plugin

Making the necessary changes to enable the new pytest plugin.
By default Twister should work without the pytest-twister-harness
plugin installed. To achieve this, each time Twister calls pytest,
the PYTHONPATH environment variable is expanded and the
`-p twister_harness.plugin` option is added to the pytest command.

Co-authored-by: Piotr Golyzniak <piotr.golyzniak@nordicsemi.no>
Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

PerMac commented:
@nashif we plan to squash all commits from this PR and leave 3: add pytest plugin, add docs, and add example of pytest test. What do you think about it?

gchwier committed:
samples: testsuite: exemplary pytest shell test

Adding exemplary pytest shell test to show possibilities of new pytest
plugin. This test uses bidirectional communication between tester and
device under test.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

PerMac committed:
ci: twister: add unit tests for pytest plugin

Adding a stage to twister unit test workflow where tests for
pytest-twister-harness plugin are executed.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac committed:
doc: twister: add documentation for pytest plugin

Adding documentation for integration of twister with pytest plugin.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

SeppoTakalo commented:
This fails to run
```
twister -T samples/subsys/testsuite/pytest/shell -vv -p native_posix
...
DEBUG   - PYTEST: self._check_conflict(action)
DEBUG   - PYTEST: File "/usr/lib/python3.8/argparse.py", line 1551, in _check_conflict
DEBUG   - PYTEST: conflict_handler(action, confl_optionals)
DEBUG   - PYTEST: File "/usr/lib/python3.8/argparse.py", line 1560, in _handle_conflict_error
DEBUG   - PYTEST: raise ArgumentError(action, message % conflict_string)
DEBUG   - PYTEST: argparse.ArgumentError: argument --platform: conflicting option string: --platform
```

Am I missing something?

PerMac commented:
@SeppoTakalo Maybe you still have twister v2 istalled in your env and it is generating args conflict?

gopiotr commented:
> This fails to run
> 
> ```
> twister -T samples/subsys/testsuite/pytest/shell -vv -p native_posix
> ...
> DEBUG   - PYTEST: self._check_conflict(action)
> DEBUG   - PYTEST: File "/usr/lib/python3.8/argparse.py", line 1551, in _check_conflict
> DEBUG   - PYTEST: conflict_handler(action, confl_optionals)
> DEBUG   - PYTEST: File "/usr/lib/python3.8/argparse.py", line 1560, in _handle_conflict_error
> DEBUG   - PYTEST: raise ArgumentError(action, message % conflict_string)
> DEBUG   - PYTEST: argparse.ArgumentError: argument --platform: conflicting option string: --platform
> ```
> 
> Am I missing something?

@SeppoTakalo probably you have installed `pytest-twister` package in your environment. Please uninstall it by `pip uninstall pytest-twister` or use virtualenv.

SeppoTakalo commented:
My mistake..
I think I was accidentally launching some old Twister code.

I removed my Python virtual environment, installed it again, and used the `./scripts/twister` instead of `twister`

It works now.

PerMac commented:
@hakehuang @teburd @nashif We believe we've addressed all the comments. @teburd I think the only difference from your previous ack is that now it works by default without any installation (using installed version is even blocked without an extra flag). Can you please do a re-review?

nashif commented:
what was the reason for the pytest failing? ( i have seen this being rerun).

nashif commented:
> @nashif However, the configuration is built only once and the image is reused. The dut fixture has a flashing in its setup. It is to support the test isolation principle. It is always possible to have more asserts in a single test if you don't want reflashing

Ok, but I think this needs to be optional and something that the test decides on. Some tests might need that, some might not.

PerMac commented:
> what was the reason for the pytest failing? ( i have seen this being rerun).

Some instability during a communication with qemu. We think it was due to a slower qemu performance. We plan to add stress/performance tests to evaluate the stability and see how it can be improved.
